File: src/org/serviceProvider/services/MicroserviceInstance.java
package org.serviceProvider.services;

import org.enduser.networkPacket.NetworkCloudlet;
import org.infrastructureProvider.entities.Vm;
import org.infrastructureProvider.policies.CloudletScheduler;
import org.utils.PolicyConstants;

import java.util.ArrayList;

public class MicroserviceInstance extends Vm {

    public int serviceId;

    /* Instance status */
//    Requested = 0;
//    Started = 1;
//    Quarantined = 2; // for Instance migration or service replacement
//    Destroyed = 3;

    public int status;

    public double requestTime;
    public double startTime;
    public double destroyTime;
    public double lifeTime; // for budget

    public double delayInStartUp;

    public int configurationType;
    public int purchaseType;
    public double price;
    public double bill;

    /* for cloudlets management */
    public ArrayList<NetworkCloudlet> cloudletList;


    public double lastUpateTime; // record last cloudlets processing time

    // for service0
    public int serviceChainId; // usually do not be used


    /**
     * Creates a new VMCharacteristics object.
     *
     * @param id                unique ID of the VM
     * @param userId            ID of the VM's owner
     * @param mips              the mips
     * @param numberOfPes       amount of CPUs
     * @param ram               amount of ram
     * @param bw                amount of bandwidth
     * @param size              amount of storage
     * @param vmm               virtual machine monitor
     * @param cloudletScheduler cloudletScheduler policy for cloudlets
     * @pre id >= 0
     * @pre userId >= 0
     * @pre size > 0
     * @pre ram > 0
     * @pre bw > 0
     * @pre cpus > 0
     * @pre priority >= 0
     * @pre cloudletScheduler != null
     * @post $none
     */
    public MicroserviceInstance(int id, int userId, double mips, int numberOfPes, int ram, long bw, long size, String vmm, CloudletScheduler cloudletScheduler
                                ,double requestTime,
                                int configurationType,
                                int purchaseType,
                                double delayInStartUp,
                                int serviceId) {
        super(id, userId, mips, numberOfPes, ram, bw, size, vmm, cloudletScheduler);
        this.requestTime = requestTime;
        this.configurationType = configurationType;
        this.purchaseType = purchaseType;
        this.delayInStartUp = delayInStartUp;
        this.serviceId = serviceId;
        this.startTime = -1;
        this.destroyTime = -1;
        this.lifeTime = 0;
        this.price = PolicyConstants.VM_PRICE[purchaseType][configurationType];
        this.bill = 0;

        setStatus(-1);
        this.cloudletList = new ArrayList<>();
        lastUpateTime = 0;

    }


    public int getServiceId() {
        return serviceId;
    }

    public void setServiceId(int serviceId) {
        this.serviceId = serviceId;
    }

    public int getStatus() {
        return status;
    }

    public void setStatus(int status) {
        this.status = status;
    }

    public double getRequestTime() {
        return requestTime;
    }

    public void setRequestTime(double requestTime) {
        this.requestTime = requestTime;
    }

    public double getStartTime() {
        return startTime;
    }

    public void setStartTime(double startTime) {
        this.startTime = startTime;
    }

    public double getDestroyTime() {
        return destroyTime;
    }

    public void setDestroyTime(double destroyTime) {
        this.destroyTime = destroyTime;
    }

    public double getLifeTime() {
        return lifeTime;
    }

    public void setLifeTime(double lifeTime) {
        this.lifeTime = lifeTime;
    }

    public double getDelayInStartUp() {
        return delayInStartUp;
    }

    public void setDelayInStartUp(double delayInStartUp) {
        this.delayInStartUp = delayInStartUp;
    }

    public int getConfigurationType() {
        return configurationType;
    }

    public void setConfigurationType(int configurationType) {
        this.configurationType = configurationType;
    }

    public int getPurchaseType() {
        return purchaseType;
    }

    public void setPurchaseType(int purchaseType) {
        this.purchaseType = purchaseType;
    }

    public double getPrice() {
        return price;
    }

    public void setPrice(double price) {
        this.price = price;
    }

    public double getBill() {
        return bill;
    }

    public void setBill(double bill) {
        this.bill = bill;
    }

    public ArrayList<NetworkCloudlet> getCloudletList() {
        return cloudletList;
    }

    public void setCloudletList(ArrayList<NetworkCloudlet> cloudletList) {
        this.cloudletList = cloudletList;
    }

    public double getLastUpateTime() {
        return lastUpateTime;
    }

    public void setLastUpateTime(double lastUpateTime) {
        this.lastUpateTime = lastUpateTime;
    }
}

File: components/ServiceDiscovery/micro_service_instance.py
class MicroserviceInstance:
    VM_PRICE = {
            {1, 1, 1, 1, 1},
            {2, 2, 2, 2, 1}
    }
    def __init__(self, id, user_id, mips, number_of_pes, ram, bw, size, vmm, cloudlet_scheduler, 
                 request_time=0.0, configuration_type=0, purchase_type=0, delay_in_startup=0.0, service_id=0):
        # Initializing attributes
        self.id = id
        self.user_id = user_id
        self.mips = mips
        self.number_of_pes = number_of_pes
        self.ram = ram
        self.bw = bw
        self.size = size
        self.vmm = vmm
        self.cloudlet_scheduler = cloudlet_scheduler
        self.request_time = request_time
        self.configuration_type = configuration_type
        self.purchase_type = purchase_type
        self.delay_in_startup = delay_in_startup
        self.service_id = service_id

        # Default attributes
        self.start_time = -1.0
        self.destroy_time = -1.0
        self.life_time = 0.0
        self.price = MicroserviceInstance.VM_PRICE[purchase_type][configuration_type]
        self.bill = 0.0
        self.status = -1  # Default status
        self.cloudlet_list = []
        self.last_update_time = 0.0

    # Getter and Setter methods
    def get_service_id(self):
        return self.service_id

    def set_service_id(self, service_id):
        self.service_id = service_id

    def get_status(self):
        return self.status

    def set_status(self, status):
        self.status = status

    def get_request_time(self):
        return self.request_time

    def set_request_time(self, request_time):
        self.request_time = request_time

    def get_start_time(self):
        return self.start_time

    def set_start_time(self, start_time):
        self.start_time = start_time

    def get_destroy_time(self):
        return self.destroy_time

    def set_destroy_time(self, destroy_time):
        self.destroy_time = destroy_time

    def get_life_time(self):
        return self.life_time

    def set_life_time(self, life_time):
        self.life_time = life_time

    def get_delay_in_startup(self):
        return self.delay_in_startup

    def set_delay_in_startup(self, delay_in_startup):
        self.delay_in_startup = delay_in_startup

    def get_configuration_type(self):
        return self.configuration_type

    def set_configuration_type(self, configuration_type):
        self.configuration_type = configuration_type

    def get_purchase_type(self):
        return self.purchase_type

    def set_purchase_type(self, purchase_type):
        self.purchase_type = purchase_type

    def get_price(self):
        return self.price  

    def set_price(self, price):
        self.price = price

    def get_bill(self):
        return self.bill

    def set_bill(self, bill):
        self.bill = bill

    def get_cloudlet_list(self):
        return self.cloudlet_list

    def set_cloudlet_list(self, cloudlet_list):
        self.cloudlet_list = cloudlet_list

    def get_last_update_time(self):
        return self.last_update_time

    def set_last_update_time(self, last_update_time):
        self.last_update_time = last_update_time
